### DISCUSSION TITLE
What is a better way to specify a type for the transform option in useFetch?

### DISCUSSION THREAD

**Discussion by suke** [nuxt/nuxt#16929] [Questions]
Hi there. 
A type error occurs when the transform option of useFetch is specified, as in the following code.

```typescript
type Mountain = {
  continent: string
  countries: string[]
  description: string
  dir: string
  height: string
  image: string
  path: string
  slug: string
  title: string
  updatedAt: string
}

const { data } = useFetch<Mountain[]>('https://api.nuxtjs.dev/mountains', {
  transform: res => res.map(mountain => mountain.slug)
})
```

<img width="590" alt="Screen Shot 2022-08-31 at 23 18 04" src="https://user-images.githubusercontent.com/38808706/187701234-26118b4b-efa5-44b8-9ff0-acfdb0e1ec23.png">

The error can be resolved by passing the type of the transform to the generics of useFetch, but this seems redundant.

```typescript
const transform = (res: Mountain[]) => res.map(mountain => mountain.slug)
const { data } = useFetch<Mountain[], Error, string, Mountain[], typeof transform>(
  'https://api.nuxtjs.dev/mountains',
  {
    transform,
  }
)
```

<img width="608" alt="Screen Shot 2022-08-31 at 23 21 22" src="https://user-images.githubusercontent.com/38808706/187701965-83f92a05-b85a-4aa6-afc6-b78157e55dfe.png">

Do you have any good suggestions?

**Comment by palamarchukser** [comment]
this not working?

```
const { data } = useFetch<Mountain[]>('https://api.nuxtjs.dev/mountains', {
  transform: (res: Mountain[]) => res.map(mountain => mountain.slug)
})
```

**Reply by suke** [reply]
Thanks for the reply!
Unfortunately it doesn't seem to work :sob:


<img width="570" alt="Screen Shot 2022-09-04 at 22 22 38" src="https://user-images.githubusercontent.com/38808706/188316239-82fe9864-abb1-44de-9fbb-de8c7b6ec79f.png">

**Comment by Flanker72** [comment]
Seems you are trying to do something not designed by the `transform` approach.
Transform allows you to modify data regarding the given structure (you should return correct `Mountain[]` type in your case).

To get the working solution, you may try this instead:
```ts
const { data } = useFetch<string[]>('https://api.nuxtjs.dev/mountains', {
  transform: res => res.map(mountain => mountain.slug)
})
```

`useFetch` type shows the data which will be returned as the result. If you are transforming the received data to modify the returned value, you should say what value you are expecting to get instead.

**Reply by some-user123** [reply]
That doesn't work either. Typescript will complain that `slug` does not exist on type `string`. To me `transform` seems pretty useless in the current form as it requires either
- type to be exactly the same (not what I consider transforming)
- overly complex type definition as mentioned by @suke (that could be inferred from the transform function)

@danielroe Would you consider this a bug?

**Reply by suke** [reply]
Yes, it does not work.
The first type parameter of useFetch is the type of the response from the API.
Transform takes it as an argument and returns the converted type.

However, since the return type of transform currently expects the API response type by default, the only way is to explicitly specify the transform type in useFetch.
